pkg/networkservice/common/contexttranslation/client.go

```diff
@@ -0,0 +1,72 @@
+// Copyright (c) 2020-2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package contexttranslation provides client chain element to perform serverRequest -> clientRequest and
+// clientConn -> serverConn context translations
+package contexttranslation
+
+import (
+	"context"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"google.golang.org/grpc"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+
+	"github.com/networkservicemesh/sdk/pkg/networkservice/core/next"
+)
+
+type contextTranslationClient struct {
+	connContexts contextMap
+}
+
+// NewClient returns a new translation client chain element
+func NewClient() networkservice.NetworkServiceClient {
+	return new(contextTranslationClient)
+}
+
+func (c *contextTranslationClient) Request(ctx context.Context, request *networkservice.NetworkServiceRequest, opts ...grpc.CallOption) (conn *networkservice.Connection, err error) {
+	connID := request.GetConnection().GetId()
+
+	// 1. Translate request Contexts
+	clientRequest := request.Clone()
+
+	connCtx, _ := c.connContexts.LoadOrStore(connID, &networkservice.ConnectionContext{})
+	clientRequest.Connection.Context = connCtx
+
+	// 2. Request client chain
+	clientConn, err := next.Client(ctx).Request(ctx, clientRequest, opts...)
+	if err != nil {
+		return nil, err
+	}
+	c.connContexts.Store(connID, clientConn.Context)
+
+	// 3. Translate connection Context
+	conn = clientConn.Clone()
+	conn.Context = request.Connection.GetContext()
+
+	return conn, nil
+}
+
+func (c *contextTranslationClient) Close(ctx context.Context, conn *networkservice.Connection, opts ...grpc.CallOption) (*empty.Empty, error) {
+	// 1. Translate connection Context
+	conn = conn.Clone()
+	connContext, _ := c.connContexts.LoadAndDelete(conn.GetId())
+	conn.Context = connContext
+
+	// 2. Close client chain
+	return next.Client(ctx).Close(ctx, conn, opts...)
```

Comment:
I see that in `Request` we restore original conn context before returning.
Shouldn't we also restore conn context in `Close`?